Panic when creating a MT

Looks like dvote/db should check for the type before converting the interface but that means arbo is trying to write a transaction of wrong type.

```
panic: interface conversion: db.WriteTx is *prefixeddb.PrefixedWriteTx, not pebbledb.WriteTx

goroutine 486974 [running]:
go.vocdoni.io/dvote/db/pebbledb.WriteTx.Apply({0xc00486f980}, {0x31bebb0, 0xc00c6ca3c0})
        /home/vocdoni/go-dvote/db/pebbledb/pebledb.go:72 +0x54
go.vocdoni.io/dvote/db.(*Batch).Apply(0x80, {0x31bebb0, 0xc00c6ca3c0})
        /home/vocdoni/go-dvote/db/batch.go:61 +0x2b
go.vocdoni.io/dvote/db/prefixeddb.(*PrefixedWriteTx).Apply(0xc011c28ff0, {0x31bebb0, 0xc00c6ca3c0})
        /home/vocdoni/go-dvote/db/prefixeddb/prefixeddb.go:133 +0x2c
github.com/vocdoni/arbo.(*Tree).addBatchInDisk(0xc0144a0770, {0x31bebb0, 0xc00c6ca3c0}, {0xc002023800, 0xc017157218, 0x13276f4}, {0xc002024400, 0x0, 0x734c})
        /home/vocdoni/go/pkg/mod/github.com/vocdoni/arbo@v0.0.0-20211217085703-d56ab859f109/tree.go:315 +0x499
github.com/vocdoni/arbo.(*Tree).AddBatchWithTx(0xc0144a0770, {0x31bebb0, 0xc00c6ca3c0}, {0xc002023800, 0x64, 0x80}, {0xc002024400, 0xc0171572a8, 0xe6e2ad})
        /home/vocdoni/go/pkg/mod/github.com/vocdoni/arbo@v0.0.0-20211217085703-d56ab859f109/tree.go:241 +0x269
go.vocdoni.io/dvote/tree.(*Tree).AddBatch(0xc0133e1488, {0x0, 0x0}, {0xc002023800, 0x64, 0x80}, {0xc002024400, 0x64, 0x80})
        /home/vocdoni/go-dvote/tree/tree.go:132 +0x170
go.vocdoni.io/dvote/censustree.(*Tree).AddBatch(0x20, {0xc002023800, 0xc013fc6580, 0x40}, {0xc002024400, 0x2d18dfe, 0x15})
        /home/vocdoni/go-dvote/censustree/censustree.go:149 +0x146
go.vocdoni.io/dvote/census.(*Manager).Handler(0xc002766460, {0x31abcc8, 0xc005759860}, 0xc01a6d1b80, {0xc01344bd70, 0xc002e99ed8})
        /home/vocdoni/go-dvote/census/handler.go:135 +0x2c46
go.vocdoni.io/dvote/rpcapi.(*RPCAPI).censusLocal(0xc001420360, 0xc01a6d1b80)
        /home/vocdoni/go-dvote/rpcapi/censushandlers.go:15 +0x185
go.vocdoni.io/dvote/rpcapi.(*RPCAPI).route(0xc001420360, {{0x27cc720, 0xc016602af0}, {0xc074e24a43586b0d, 0x6bc042d01d, 0x4575f20}, {0xc00afe1070, 0x1, 0x1}, 0xc00afe1040})
        /home/vocdoni/go-dvote/rpcapi/api.go:109 +0xca
created by go.vocdoni.io/dvote/httprouter.(*HTTProuter).routerHandler.func1
        /home/vocdoni/go-dvote/httprouter/httprouter.go:241 +0x3f9
```

--- a/db/badgerdb/badger.go
+++ b/db/badgerdb/badger.go
@@ -79,7 +79,7 @@ func (tx WriteTx) Delete(k []byte) error {
 
 // Apply implements the db.WriteTx.Apply interface method
 func (tx WriteTx) Apply(other db.WriteTx) (err error) {
-	otherBadger := other.(WriteTx)
+	otherBadger := db.UnwrapWriteTx(other).(WriteTx)
 
 	valueOfTx := reflect.ValueOf(otherBadger.tx)
 	pendingWrites := valueOfTx.Elem().FieldByName("pendingWrites")

--- a/db/badgerdb/badger_test.go
+++ b/db/badgerdb/badger_test.go
@@ -7,6 +7,7 @@ import (
 	qt "github.com/frankban/quicktest"
 	"go.vocdoni.io/dvote/db"
 	"go.vocdoni.io/dvote/db/internal/dbtest"
+	"go.vocdoni.io/dvote/db/prefixeddb"
 )
 
 func TestWriteTx(t *testing.T) {
@@ -35,6 +36,23 @@ func TestWriteTxApply(t *testing.T) {
 	qt.Assert(t, err, qt.IsNil)
 
 	dbtest.TestWriteTxApply(t, database)
+}
+
+func TestWriteTxApplyPrefixed(t *testing.T) {
+	database, err := New(db.Options{Path: t.TempDir()})
+	qt.Assert(t, err, qt.IsNil)
+
+	prefix := []byte("one")
+	dbWithPrefix := prefixeddb.NewPrefixedDatabase(database, prefix)
+
+	dbtest.TestWriteTxApplyPrefixed(t, database, dbWithPrefix)
+}
+
+func TestWriteTxApplyBatch(t *testing.T) {
+	database, err := New(db.Options{Path: t.TempDir()})
+	qt.Assert(t, err, qt.IsNil)
+
+	dbtest.TestWriteTxApplyBatch(t, database)
 }
 
 func TestBatch(t *testing.T) {

--- a/db/batch.go
+++ b/db/batch.go
@@ -61,6 +61,11 @@ func (t *Batch) Apply(other WriteTx) (err error) {
 	return t.tx.Apply(other)
 }
 
+// Unwrap returns the wrapped WriteTx
+func (t *Batch) Unwrap() WriteTx {
+	return t.tx
+}
+
 // Set implements the WriteTx.Set interface method.  If during this
 // operation, the internal tx becomes too big, all the pending writes will be
 // committed and a new WriteTx will be created to continue with this and

--- a/db/interface.go
+++ b/db/interface.go
@@ -77,3 +77,14 @@ type WriteTx interface {
 	// Commit commits the transaction into the db
 	Commit() error
 }
+
+// UnwrapWriteTx unwraps (if possible) the WriteTx using Unwrap method
+func UnwrapWriteTx(tx WriteTx) WriteTx {
+	for {
+		wtx, ok := tx.(interface{ Unwrap() WriteTx })
+		if !ok {
+			return tx
+		}
+		tx = wtx.Unwrap()
+	}
+}

--- a/db/internal/dbtest/dbtest.go
+++ b/db/internal/dbtest/dbtest.go
@@ -173,3 +173,39 @@ func TestWriteTxApply(t *testing.T, d db.Database) {
 	qt.Assert(t, err, qt.IsNil)
 	qt.Assert(t, gV, qt.DeepEquals, []byte(strconv.Itoa(3)))
 }
+
+func TestWriteTxApplyPrefixed(t *testing.T, d, dbWithPrefix db.Database) {
+	wTxWithPrefix := dbWithPrefix.WriteTx()
+	defer wTxWithPrefix.Discard()
+	wTx := d.WriteTx()
+	defer wTx.Discard()
+
+	qt.Assert(t, wTxWithPrefix.Set([]byte("a"), []byte("a")), qt.IsNil)
+	qt.Assert(t, wTx.Set([]byte("b"), []byte("b")), qt.IsNil)
+
+	// try to put the contents from the normal wTx into the wTxWithPrefix
+	err := wTxWithPrefix.Apply(wTx)
+	qt.Assert(t, err, qt.IsNil)
+
+	// try to put the contents from wTxWithPrefix into the normal wTx
+	err = wTx.Apply(wTxWithPrefix)
+	qt.Assert(t, err, qt.IsNil)
+}
+
+func TestWriteTxApplyBatch(t *testing.T, d db.Database) {
+	batch := db.NewBatch(d)
+	defer batch.Discard()
+	wTx := d.WriteTx()
+	defer wTx.Discard()
+
+	qt.Assert(t, batch.Set([]byte("a"), []byte("a")), qt.IsNil)
+	qt.Assert(t, wTx.Set([]byte("b"), []byte("b")), qt.IsNil)
+
+	// try to put the contents from the normal wTx into the batch
+	err := batch.Apply(wTx)
+	qt.Assert(t, err, qt.IsNil)
+
+	// try to put the contents from batch into the normal wTx
+	err = wTx.Apply(batch)
+	qt.Assert(t, err, qt.IsNil)
+}

--- a/db/pebbledb/pebledb.go
+++ b/db/pebbledb/pebledb.go
@@ -69,7 +69,7 @@ func (tx WriteTx) Delete(k []byte) error {
 
 // Apply implements the db.WriteTx.Apply interface method
 func (tx WriteTx) Apply(other db.WriteTx) (err error) {
-	otherPebble := other.(WriteTx)
+	otherPebble := db.UnwrapWriteTx(other).(WriteTx)
 	return tx.batch.Apply(otherPebble.batch, nil)
 }
 

--- a/db/pebbledb/pebledb_test.go
+++ b/db/pebbledb/pebledb_test.go
@@ -6,6 +6,7 @@ import (
 	qt "github.com/frankban/quicktest"
 	"go.vocdoni.io/dvote/db"
 	"go.vocdoni.io/dvote/db/internal/dbtest"
+	"go.vocdoni.io/dvote/db/prefixeddb"
 )
 
 func TestWriteTx(t *testing.T) {
@@ -20,6 +21,30 @@ func TestIterate(t *testing.T) {
 	qt.Assert(t, err, qt.IsNil)
 
 	dbtest.TestIterate(t, database)
+}
+
+func TestWriteTxApply(t *testing.T) {
+	database, err := New(db.Options{Path: t.TempDir()})
+	qt.Assert(t, err, qt.IsNil)
+
+	dbtest.TestWriteTxApply(t, database)
+}
+
+func TestWriteTxApplyPrefixed(t *testing.T) {
+	database, err := New(db.Options{Path: t.TempDir()})
+	qt.Assert(t, err, qt.IsNil)
+
+	prefix := []byte("one")
+	dbWithPrefix := prefixeddb.NewPrefixedDatabase(database, prefix)
+
+	dbtest.TestWriteTxApplyPrefixed(t, database, dbWithPrefix)
+}
+
+func TestWriteTxApplyBatch(t *testing.T) {
+	database, err := New(db.Options{Path: t.TempDir()})
+	qt.Assert(t, err, qt.IsNil)
+
+	dbtest.TestWriteTxApplyBatch(t, database)
 }
 
 // NOTE: This test fails.  pebble.Batch doesn't detect conflicts.  Moreover,

--- a/db/prefixeddb/prefixeddb.go
+++ b/db/prefixeddb/prefixeddb.go
@@ -133,6 +133,11 @@ func (t *PrefixedWriteTx) Apply(other db.WriteTx) error {
 	return t.tx.Apply(other)
 }
 
+// Unwrap returns the wrapped WriteTx
+func (t *PrefixedWriteTx) Unwrap() db.WriteTx {
+	return t.tx
+}
+
 // Commit implements the db.WriteTx.Commit interface method.  Notice that this
 // method also commits the wrapped db.WriteTx.
 func (t *PrefixedWriteTx) Commit() error {

